### ISSUE TITLE
stable branch broken because of pkg_resources

### ISSUE THREAD

**Issue by ilanschnell** [biolab/orange3#1007]
I'm getting the following error on the `stable` branch:
```
$ ./orange-canvas 
Traceback (most recent call last):
  File "./orange-canvas", line 6, in <module>
    sys.exit(Orange.canvas.__main__.main())
  File "/Users/ilan/python/envs/foo/lib/python3.5/site-packages/Orange/canvas/__main__.py", line 231, in main
    if pkg_resources.resource_exists(pkg_name, resource):
  File "/Users/ilan/python/envs/foo/lib/python3.5/site-packages/setuptools-19.6.2-py3.5.egg/pkg_resources/__init__.py", line 1139, in resource_exists
  File "/Users/ilan/python/envs/foo/lib/python3.5/site-packages/setuptools-19.6.2-py3.5.egg/pkg_resources/__init__.py", line 1605, in has_resource
  File "/Users/ilan/python/envs/foo/lib/python3.5/site-packages/setuptools-19.6.2-py3.5.egg/pkg_resources/__init__.py", line 1660, in _has
NotImplementedError: Can't perform this operation for unregistered loader type
```
I recommend not using `pkg_resources` at all.  Being part of `setuptools`, I find `pkg_resources` being extremely unreliable, as it depends on strange setuptools magic.

**Comment by kernc** [comment]
> I recommend not using pkg_resources at all.

Thanks. Is there an alternative?

**Comment by ilanschnell** [comment]
That depends on the particular use-case, I would say.  One could (as an example) have something like:
```
try:
    import other_package
    resource_file = os.path.join(other_package.__file__, '..', 'style.css')
    open(resource_file).read()
except (ImportError, IOError):
    resource_file = None
```

**Comment by kernc** [comment]
The example is helpful, thanks.

Is

> `pkg_resources` being extremely unreliable, as it depends on strange setuptools magic

a strictly personal opinion or is there perhaps a wider community consensus?

I don't suppose https://github.com/biolab/orange3/pull/1006 would by chance mitigate above-reported exception?

**Comment by ilanschnell** [comment]
The wider Python seems to have adopted `setuptools` by now, even though there are still many people that don't like it.  I have been in the business of Python packaging for about seven years now (first maintaining EPD at Enthought, and now Anaconda), and `setuptools` has always been a source of frustration.  By personal projects (`bitarray`, `bsdiff4`, `pycosat`) don't use `setuptools`, just plain `distutils`.  In my opinion, it is bad enough if setuptools is used for building packages, but using it at runtime is even worse.  I create many "binary" conda packages, and once such a package is created (whether it uses setuptools or not) it is done, ie. all files which have to be installed on the end users machine are laid out nicely in a tarball (and setuptools is not part of the equation anymore).  However, if setuptools is used inside the package itself (as opposed to just `setup.py`), then any problems with setuptools continue to exist on the end users machine (for example when setuptools is updated to a newer version).

**Comment by ajdapretnar** [comment]
@ilanschnell Is the issue still pertinent?

**Comment by ajdapretnar** [comment]
@ilanschnell Fix #1006 is supposed to fix this. Does it work for you?

**Comment by ajdapretnar** [comment]
Closing the issue since the fix seems to be implemented already. @ilanschnell If the problem persists, please open a new issue.
